[Typography] Cache standard font for text style.

romoore opened the pull request:
Scrolling the `MDCSelfSizingStereoCell` while using the Time Profiler showed that `+[UIFont mdc_standardFontForMaterialTextStyle:]` was consuming about 11% (iPhone 4S) of the used CPU time.  

Because the `UIFont` returned for a specific `MDCFontTextStyle` should always be the same (they are created with exactly the same `UIFontDescriptor` values), the font can be cached using the `MDCFontTextStyle` as a key. This nearly eliminates any time required for these font retrievals (excepting the first call), and increased CPU time dedicated to the main thread by 8% (iPhone 4S). 

Reduces time spent loading fonts when scrolling Collection Cells: 11% on iPhone 4S/iOS 9.3.6, 2% on iPhone 5c/iOS 10.3.3

||Original|Cached|
|--|--|--|
|iPhone 4S/iOS 9.3.5|![typography-original](https://user-images.githubusercontent.com/1753199/47525482-60af4680-d86b-11e8-8995-bd13914385da.png)|![typography-cached](https://user-images.githubusercontent.com/1753199/47525491-64db6400-d86b-11e8-9850-98e114c60724.png)|
|iPhone 5c/iOS 10.3.3|![typo-orig-min-5c](https://user-images.githubusercontent.com/1753199/47526162-f1d2ed00-d86c-11e8-9b53-6d6b018d433f.png)|![typo-cached-min-5c](https://user-images.githubusercontent.com/1753199/47526168-f5ff0a80-d86c-11e8-8061-48375b255817.png)|

Commit:
[Typography] Cache standard font for text style.

Performance improvements.

material-automation commented on components/Typography/src/UIFont+MaterialTypography.m:
```diff
@@ -37,20 +37,45 @@ + (UIFont *)mdc_preferredFontForMaterialTextStyle:(MDCFontTextStyle)style {
 }
 
 + (nonnull UIFont *)mdc_standardFontForMaterialTextStyle:(MDCFontTextStyle)style {
+
```
clang-format suggested the following change:


```
-
```

<!-- identifier: clang-format -->
<!-- File: components/Typography/src/UIFont+MaterialTypography.m -->

romoore replied:
I disagree.

material-automation commented on components/Typography/src/UIFont+MaterialTypography.m:
```diff
@@ -37,20 +37,45 @@ + (UIFont *)mdc_preferredFontForMaterialTextStyle:(MDCFontTextStyle)style {
+    UIFont *font = [fontCache objectForKey:@(style)];
+    if (!font) {
+      UIFontDescriptor *fontDescriptor =
       [UIFontDescriptor mdc_standardFontDescriptorForMaterialTextStyle:style];
```
clang-format suggested the following change:


```
-      [UIFontDescriptor mdc_standardFontDescriptorForMaterialTextStyle:style];
+          [UIFontDescriptor mdc_standardFontDescriptorForMaterialTextStyle:style];
```

<!-- identifier: clang-format -->
<!-- File: components/Typography/src/UIFont+MaterialTypography.m -->

material-automation commented on components/Typography/src/UIFont+MaterialTypography.m:
```diff
@@ -37,20 +37,45 @@ + (UIFont *)mdc_preferredFontForMaterialTextStyle:(MDCFontTextStyle)style {
+    return font;
+  }
+  UIFontDescriptor *fontDescriptor =
+  [UIFontDescriptor mdc_standardFontDescriptorForMaterialTextStyle:style];
```
clang-format suggested the following change:


```
-  [UIFontDescriptor mdc_standardFontDescriptorForMaterialTextStyle:style];
+      [UIFontDescriptor mdc_standardFontDescriptorForMaterialTextStyle:style];
```

<!-- identifier: clang-format -->
<!-- File: components/Typography/src/UIFont+MaterialTypography.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Adding unit test.

material-automation commented on components/Typography/tests/unit/UIFont+MaterialTypographyTests.m:
```diff
@@ -0,0 +1,47 @@
+      @[@(MDCFontTextStyleTitle), @(MDCFontTextStyleBody1), @(MDCFontTextStyleBody2),
+        @(MDCFontTextStyleButton), @(MDCFontTextStyleCaption), @(MDCFontTextStyleDisplay1),
+        @(MDCFontTextStyleDisplay2), @(MDCFontTextStyleDisplay3), @(MDCFontTextStyleDisplay4),
+        @(MDCFontTextStyleHeadline), @(MDCFontTextStyleSubheadline)];
```
clang-format suggested the following change:


```
-  NSArray<NSNumber *> *textStyles =
-      @[@(MDCFontTextStyleTitle), @(MDCFontTextStyleBody1), @(MDCFontTextStyleBody2),
-        @(MDCFontTextStyleButton), @(MDCFontTextStyleCaption), @(MDCFontTextStyleDisplay1),
-        @(MDCFontTextStyleDisplay2), @(MDCFontTextStyleDisplay3), @(MDCFontTextStyleDisplay4),
-        @(MDCFontTextStyleHeadline), @(MDCFontTextStyleSubheadline)];
+  NSArray<NSNumber *> *textStyles = @[
+    @(MDCFontTextStyleTitle), @(MDCFontTextStyleBody1), @(MDCFontTextStyleBody2),
+    @(MDCFontTextStyleButton), @(MDCFontTextStyleCaption), @(MDCFontTextStyleDisplay1),
+    @(MDCFontTextStyleDisplay2), @(MDCFontTextStyleDisplay3), @(MDCFontTextStyleDisplay4),
+    @(MDCFontTextStyleHeadline), @(MDCFontTextStyleSubheadline)
+  ];
```

<!-- identifier: clang-format -->
<!-- File: components/Typography/tests/unit/UIFont+MaterialTypographyTests.m -->

romoore replied:
Done

Commit:
Fixing array format.

ianegordon reviewed (approved):
The approach LGTM.
Can we do a global presubmit against internal clients?
If that looks good, and we can sort out our bazel checks, please merge.

Commit:
Remove test that checks every font for uniqueness.

Originally, the test included a "sanity check" that no two subsequent
text styles produced equivalent UIFont objects. This assumption was
wrong and the test failed.

romoore commented:
Internal tests showed no failures related to this change.
